File: bootstrap/python/_parser.py
from _lexer import *

__all__: list[str] = ["Parser",]

class Parser:
    def __init__(self, lexer: Lexer) -> None:
        self.lexer: Lexer = lexer
        self.last_token: Token | None = None
        self.curr_token: Token | None = None
        self.next_token: Token | None = None
        
        self.program: list[tuple[int]] = []

        # populate self.curr_token and self.next_token
        self.next()
        self.next()

    def next(self) -> None:
        self.last_token = self.curr_token
        self.curr_token = self.next_token
        self.next_token = self.lexer.next_token()

    def try_integer(self, token: Token) -> tuple[int] | None:
        # (type, value)
        if token.kind != TokenKind.digit:
            assert False, f"ERROR:{token.start.line}:{token.start.column}: try_integer() expected token of kind `digit` ({TokenKind.digit})!"
        if token.end.offset-token.start.offset > 20:
            assert False, f"ERROR:{token.start.line}:{token.start.column}: try_integer() length of token ({token.end.offset-token.start.offset}) seams to be over 20!"
        text: str = self.lexer.text[token.start.offset:token.end.offset]
        value: int = 0
        for i in len(text):
            value = value * 10 + int(text[i], 10)
        return (Type.i64, value)

    def try_add(self) -> tuple[int] | None:
        # (instruction, value0, value1)
        if (self.last_char.kind == TokenKind.digit and 
            self.next_char.kind == TokenKind.digit):
            return (Ins.add, self.try_integer(), self.try_integer())

    def parse(self) -> list[tuple[int]]:
        while isinstance(self.curr_token, str):
            if self.curr_token.kind == TokenKind.plus:
                self.try_add()
            else:
                print(TokenKind())
                assert False, f"ERROR:{self.curr_token.start.line}:{self.curr_token.start.column}: Encountered an invalid token ({self.curr_token.kind}) while parsing!"

File: src/c/parser.c
#include "parser.h"

parser_t *new_parser(lexer_t *lexer) {
    parser_t *parser = malloc(sizeof(parser_t));
    if (!parser) return NULL;

    parser->lexer = lexer;
    parser->curr_token = new_token (TOKEN_UNDEFINED, 
            new_span (new_token (1, 1, 0, 0), new_token (1, 1, 0, 0)));
    parser->next_token = new_token (TOKEN_UNDEFINED,
            new_span (new_token (1, 1, 0, 0), new_token (1, 1, 0, 0)));

    parser_next(parser);
    return parser;
}

void del_paresr(parser_t **parser) {
    // parser->lexer has to be manually freed
    free(*parser);
    *parser = NULL;
}

void parser_next(parser_t *parser) {
    parser->last_token = parser->curr_token;
    parser->curr_token = lexer_next_token(parser->lexer);
}

expression_t parser_expression(parser_t *parser, unsigned char priority) {
    position_t start_position = parser->curr_token.span.start;
    expression_t left;
    
    switch (parser->curr_token.kind) {
    case TOKEN_INTEGER:
        left = parser_value_expr(parser);
    }

    while (get_priority(parser->curr_token.kind).right > priority) {
        switch (parser->curr_token.kind) {
        case TOKEN_PLUS:
            left = parser_infix_expr(
                    parser, start_position, INFIX_PLUS, &left);
        }
    }

    return left;
}

int64_t parser_i64(parser_t *parser) {
    parser_next(parser);

    if (parser->last_token.kind != TOKEN_INTEGER) {
        fprintf(
            stderr, 
            "ERROR:Parser:%d:%d:%s: Encountered an invalid token "
            "while parsing i64!\n", 
            parser->last_token.span.line, 
            parser->last_token.span.column, 
            parser->lexer->path
        );
        return 0;
    }

    int64_t sign = 1;
    if (parser->last_token.kind == TOKEN_MINUS) {
        num *= -1;
    }
    
    int64_t num = 0;
    for (size_t offset = parser->last_token.span.end.offset;
            offset >= parser->last_token.span.start.offset; offset++) {
        num *= 10;
        num += parser->lexer->text[offset] - 0x30;
    }

    return num * sign;
}

expression_t parser_value_expr(parser_t *parser, position_t *start_position) {
    value_expr_t value = new_value_expr (
        TYPE_UNDEFINED,
        NULL,
    );

    switch (parser->last_token.kind) {
    case TOKEN_INTEGER:
        value.value = malloc(sizeof(int64_t));
        *value.value = parser_i64(parser);
    }

    expression_t expr = new_expression (
        EXPR_VALUE,
        new_span (start_position, parser->last_token.span.end),
        malloc(sizeof(value_expr_t))
    );

    *expr->expr = value;
    return expr;
}

expression_t parser_infix_expr(parser_t *parser, position_t start_position, 
                               infix_op_t op, expression_t *left) {
    unsigned char right_priority = get_priority(parser->curr_token.kind).right;
    parser_next(parser);
    expression_t right = parser_expression(parser, right_priority);
    
    infix_expr_t infix = new_infix_expr (
        op, 
        malloc(sizeof(expression_t)),
        malloc(sizeof(expression_t))
    );

    *infix->lvalue = *left; 
    *infix->rvalue = right;

    expression_t expr = new_expression (
        EXPR_INFIX, 
        new_span (start_position, parser->last_token.span.end),
        malloc(sizeof(infix_expr_t));
    );

    *expr->expr = infix;
    return expr;
}
